packages/client-browser/src/connection/browser_connection.ts

```diff
@@ -0,0 +1,178 @@
+      query_id,
+      stream: response.body || new ReadableStream<Uint8Array>(),
+    }
+  }
+
+  async exec(params: BaseQueryParams): Promise<QueryResult<ReadableStream>> {
+    const query_id = getQueryId(params.query_id)
+    const searchParams = toSearchParams({
+      database: this.params.database,
+      clickhouse_settings: params.clickhouse_settings,
+      query_params: params.query_params,
+      session_id: params.session_id,
+      query_id,
+    })
+    const response = await this.request({
+      body: params.query,
+      params,
+      searchParams,
+    })
+    return {
+      stream: response.body || new ReadableStream<Uint8Array>(),
+      query_id,
+    }
+  }
+
+  async insert<T = unknown>(
+    params: InsertParams<ReadableStream<T>>
+  ): Promise<InsertResult> {
+    const query_id = getQueryId(params.query_id)
+    const searchParams = toSearchParams({
+      database: this.params.database,
+      clickhouse_settings: params.clickhouse_settings,
+      query_params: params.query_params,
+      query: params.query,
+      session_id: params.session_id,
+      query_id,
+    })
+    await this.request({
+      body: params.values,
+      params,
+      searchParams,
+    })
+    return {
+      query_id,
+    }
+  }
+
+  async ping(): Promise<boolean> {
+    // TODO: catch an error and just log it, returning false?
+    const response = await this.request({
+      method: 'GET',
+      body: null,
+      pathname: '/ping',
+      searchParams: undefined,
+    })
+    if (response.body !== null) {
+      await response.body.cancel()
+    }
+    return true
+  }
+
+  async close(): Promise<void> {
+    return
+  }
+
+  private async request({
+    body,
+    params,
+    searchParams,
+    pathname,
+    method,
+  }: {
+    body: string | ReadableStream | null
+    params?: BaseQueryParams
+    searchParams: URLSearchParams | undefined
+    pathname?: string
+    method?: 'GET' | 'POST'
+  }): Promise<Response> {
+    const url = transformUrl({
+      url: this.params.url,
+      pathname: pathname ?? '/',
+      searchParams,
+    }).toString()
+    const abortController = params?.abort_controller || new AbortController()
+    let isTimedOut = false
+    const timeout = setTimeout(() => {
+      isTimedOut = true
+      abortController.abort('Request timed out')
+    }, this.params.request_timeout)
+    const bodyIsStream = isStream(body)
+    try {
+      const response = await fetch(url, {
+        body,
+        keepalive: !bodyIsStream,
+        method: method ?? 'POST',
+        signal: abortController.signal,
+        headers: withCompressionHeaders({
+          headers: this.defaultHeaders,
+          compress_request: false,
+          decompress_response: this.params.compression.decompress_response,
+        }),
+        // @ts-expect-error 'duplex' does not exist in type 'RequestInit'
+        duplex: bodyIsStream ? 'half' : undefined, // https://developer.chrome.com/articles/fetch-streaming-requests/
```

Comment:
Does it work on firefox? Does it even work with chrome? No, according to the comment https://github.com/ClickHouse/clickhouse-js/pull/165/files#diff-2c9af14af89d1014d64398505017be35255b65ad8a383b0228dee69c9f6eabdeR6-R8
Maybe we shouldn't support `insert` streaming?

Reply:
Nested/dependent generics were not trivial to implement for this scenario. I ended up just restricting the type returned by the browser's `createClient` method.